db/src/bootstrap.rs

```diff
@@ -0,0 +1,210 @@
+            .unwrap()
+    };
+
+    static ref V2_SYMBOLIC_SCHEMA: Value = {
+        let s = r#"
+{:db.alter/attribute   {:db/valueType   :db.type/ref
+                        :db/cardinality :db.cardinality/many}
+ :db.schema/version    {:db/valueType   :db.type/long
+                        :db/cardinality :db.cardinality/one}
+
+ ;; unique-value because an attribute can only belong to a single
+ ;; schema fragment.
+ :db.schema/attribute  {:db/valueType   :db.type/ref
+                        :db/unique      :db.unique/value
+                        :db/cardinality :db.cardinality/many}}"#;
+        edn::parse::value(s)
+            .map_err(|_| ErrorKind::BadBootstrapDefinition("Unable to parse V2_SYMBOLIC_SCHEMA".into()))
+            .unwrap()
+    };
+}
+
+/// Convert (ident, entid) pairs into [:db/add IDENT :db/ident IDENT] `Value` instances.
+fn idents_to_assertions(idents: &[(&str, i64)]) -> Vec<Value> {
+    idents
+        .into_iter()
+        .map(|&(ident, _)| {
+            let value = Value::NamespacedKeyword(to_namespaced_keyword(&ident).unwrap());
+            Value::Vector(vec![values::DB_ADD.clone(), value.clone(), values::DB_IDENT.clone(), value.clone()])
+        })
+        .collect()
+}
+
+/// Convert {IDENT {:key :value ...} ...} to [[:db/add IDENT :key :value] ...].
+/// In addition, add [:db.add :db.part/db :db.install/attribute IDENT] installation assertions.
+fn symbolic_schema_to_assertions(symbolic_schema: &Value) -> Result<Vec<Value>> {
+    // Failure here is a coding error, not a runtime error.
+    let mut assertions: Vec<Value> = vec![];
+    match *symbolic_schema {
+        Value::Map(ref m) => {
+            for (ident, mp) in m {
+                assertions.push(Value::Vector(vec![values::DB_ADD.clone(),
+                                                   values::DB_PART_DB.clone(),
+                                                   values::DB_INSTALL_ATTRIBUTE.clone(),
+                                                   ident.clone()]));
+                match *mp {
+                    Value::Map(ref mpp) => {
+                        for (attr, value) in mpp {
+                            assertions.push(Value::Vector(vec![values::DB_ADD.clone(),
+                                                               ident.clone(),
+                                                               attr.clone(),
+                                                               value.clone()]));
+                        }
+                    },
+                    _ => bail!(ErrorKind::BadBootstrapDefinition("Expected {:db/ident {:db/attr value ...} ...}".into()))
+                }
+            }
+        },
+        _ => bail!(ErrorKind::BadBootstrapDefinition("Expected {...}".into()))
+    }
+    Ok(assertions)
+}
+
+pub fn bootstrap_partition_map() -> PartitionMap {
+    V1_PARTS[..].into_iter()
+        .chain(V2_PARTS[..].into_iter())
```

Comment:
Sure, I'll duplicate this.  I had it that way originally, and did this to be similar to what happened elsewhere; I'll instead decomplect `V1` from `V2` in the code, and make the variables encode the extensions.  (Assuming I can figure out the correct expression of that idea.)